File: library/src/test/java/me/proxer/library/api/NumberBasedBooleanAdapterTest.java
package me.proxer.library.api;

import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.JsonDataException;
import me.proxer.library.ProxerTest;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.security.GeneralSecurityException;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;

/**
 * @author Ruben Gees
 */
class NumberBasedBooleanAdapterTest extends ProxerTest {

    private JsonAdapter<Boolean> adapter;

    @Override
    @BeforeEach
    protected void setUp() throws IOException, GeneralSecurityException {
        super.setUp();

        adapter = api.moshi().adapter(Boolean.class, NumberBasedBoolean.class);
    }

    @SuppressWarnings({"CheckReturnValue", "ResultOfMethodCallIgnored"})
    @Test
    void testFromBoolean() {
        assertThatExceptionOfType(JsonDataException.class)
                .isThrownBy(() -> adapter.fromJson("true"))
                .withMessage("Expected a number, string or null but was BOOLEAN at path $");
    }

    @SuppressWarnings({"CheckReturnValue", "ResultOfMethodCallIgnored"})
    @Test
    void testFromBooleanFalse() {
        assertThatExceptionOfType(JsonDataException.class)
                .isThrownBy(() -> adapter.fromJson("false"))
                .withMessage("Expected a number, string or null but was BOOLEAN at path $");
    }

    @Test
    void testFromInt() throws IOException {
        assertThat(adapter.fromJson("1")).isTrue();
    }

    @Test
    void testFromIntFalse() throws IOException {
        assertThat(adapter.fromJson("0")).isFalse();
    }

    @SuppressWarnings({"CheckReturnValue", "ResultOfMethodCallIgnored"})
    @Test
    void testFromIntInvalid() {
        assertThatExceptionOfType(JsonDataException.class)
                .isThrownBy(() -> adapter.fromJson("3"))
                .withMessage("Expected a 1 or 0 but was 3 at path $");
    }

    @Test
    void testFromString() throws IOException {
        assertThat(adapter.fromJson("\"true\"")).isTrue();
    }

    @Test
    void testFromStringFalse() throws IOException {
        assertThat(adapter.fromJson("\"false\"")).isFalse();
    }

    @SuppressWarnings({"CheckReturnValue", "ResultOfMethodCallIgnored"})
    @Test
    void testFromStringInvalid() {
        assertThatExceptionOfType(JsonDataException.class)
                .isThrownBy(() -> adapter.fromJson("\"invalid\""))
                .withMessage("Expected a true or false but was invalid at path $");
    }

    @Test
    void testFromStringInt() throws IOException {
        assertThat(adapter.fromJson("\"1\"")).isTrue();
    }

    @Test
    void testFromStringIntFalse() throws IOException {
        assertThat(adapter.fromJson("\"0\"")).isFalse();
    }

    @SuppressWarnings({"CheckReturnValue", "ResultOfMethodCallIgnored"})
    @Test
    void testFromStringIntInvalid() {
        assertThatExceptionOfType(JsonDataException.class)
                .isThrownBy(() -> adapter.fromJson("\"3\""))
                .withMessage("Expected a 1 or 0 but was 3 at path $");
    }

    @Test
    void testFromNull() throws IOException {
        assertThat(adapter.fromJson("null")).isNull();
    }

    @SuppressWarnings({"CheckReturnValue", "ResultOfMethodCallIgnored"})
    @Test
    void testFromInvalid() {
        assertThatExceptionOfType(JsonDataException.class)
                .isThrownBy(() -> adapter.fromJson("[]"))
                .withMessage("Expected a number, string or null but was BEGIN_ARRAY at path $");
    }

    @Test
    void testToJson() {
        assertThat(adapter.toJson(true)).isEqualTo("1");
    }

    @Test
    void testToJsonFalse() {
        assertThat(adapter.toJson(false)).isEqualTo("0");
    }

    @Test
    void testToJsonNull() {
        assertThat(adapter.toJson(null)).isEqualTo("null");
    }
}

File: library/src/test/kotlin/me/proxer/library/internal/adapter/NumberBasedBooleanAdapterTest.kt
package me.proxer.library.internal.adapter

import com.squareup.moshi.JsonDataException
import com.squareup.moshi.Moshi
import org.amshove.kluent.invoking
import org.amshove.kluent.shouldBe
import org.amshove.kluent.shouldBeNull
import org.amshove.kluent.shouldEqual
import org.amshove.kluent.shouldThrow
import org.junit.jupiter.api.Test

/**
 * @author Ruben Gees
 */
class NumberBasedBooleanAdapterTest {

    private val adapter = Moshi.Builder()
        .add(BooleanAdapterFactory())
        .build()
        .adapter<Boolean>(Boolean::class.java, NumberBasedBoolean::class.java)

    @Test
    fun testFromBoolean() {
        val result = invoking { adapter.fromJson("true") } shouldThrow JsonDataException::class

        result.exceptionMessage shouldEqual "Expected a number, string or null but was BOOLEAN at path $"
    }

    @Test
    fun testFromBooleanFalse() {
        val result = invoking { adapter.fromJson("false") } shouldThrow JsonDataException::class

        result.exceptionMessage shouldEqual "Expected a number, string or null but was BOOLEAN at path $"
    }

    @Test
    fun testFromInt() {
        adapter.fromJson("1") shouldBe true
    }

    @Test
    fun testFromIntFalse() {
        adapter.fromJson("0") shouldBe false
    }

    @Test
    fun testFromIntInvalid() {
        val result = invoking { adapter.fromJson("3") } shouldThrow JsonDataException::class

        result.exceptionMessage shouldEqual "Expected a 1 or 0 but was 3 at path $"
    }

    @Test
    fun testFromString() {
        adapter.fromJson(""""true"""") shouldBe true
    }

    @Test
    fun testFromStringFalse() {
        adapter.fromJson(""""false"""") shouldBe false
    }

    @Test
    fun testFromStringInvalid() {
        val result = invoking { adapter.fromJson(""""invalid"""") } shouldThrow JsonDataException::class

        result.exceptionMessage shouldEqual "Expected a true or false but was invalid at path $"
    }

    @Test
    fun testFromStringInt() {
        adapter.fromJson(""""1"""") shouldBe true
    }

    @Test
    fun testFromStringIntFalse() {
        adapter.fromJson(""""0"""") shouldBe false
    }

    @Test
    fun testFromStringIntInvalid() {
        val result = invoking { adapter.fromJson(""""3"""") } shouldThrow JsonDataException::class

        result.exceptionMessage shouldEqual "Expected a 1 or 0 but was 3 at path $"
    }

    @Test
    fun testFromNull() {
        adapter.fromJson("null").shouldBeNull()
    }

    @Test
    fun testFromInvalid() {
        val result = invoking { adapter.fromJson("[]") } shouldThrow JsonDataException::class

        result.exceptionMessage shouldEqual "Expected a number, string or null but was BEGIN_ARRAY at path $"
    }

    @Test
    fun testToJson() {
        adapter.toJson(true) shouldEqual "1"
    }

    @Test
    fun testToJsonFalse() {
        adapter.toJson(false) shouldEqual "0"
    }

    @Test
    fun testToJsonNull() {
        adapter.toJson(null) shouldEqual "null"
    }
}
